portfolio/src/main/webapp/game_script.js

```diff
@@ -0,0 +1,199 @@
+  canvas : document.createElement("canvas"),
+  start : function() {
+    this.canvas.width = window.innerWidth;
+    this.canvas.height = window.innerHeight;
+    this.context = this.canvas.getContext("2d");
+    document.body.insertBefore(this.canvas, document.body.childNodes[0]);
+    this.frameNo = 0;
+    this.interval = setInterval(updateGameArea, 20);
+    gameArea.mouseControl = false;
+    window.addEventListener('keydown', function (e) {
+      gameArea.key = e.keyCode;
+    })
+    window.addEventListener('keyup', function (e) {
+      if (gameArea.key == 77) {
+        /* if the key that was just pressed was "m"
+           activate mouse controls */ 
+          gameArea.mouseControl = !gameArea.mouseControl;
+      }
+      gameArea.key = false;
+    })
+    window.addEventListener('mousemove', function (e) {
+      gameArea.x = e.pageX;
+      gameArea.y = e.pageY;
+    })
+  },
+  clear : function() {
+      this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
+  },
+  stop : function() {
+    clearInterval(this.interval);
+  },
+  updateInstructions : function() {
+      INSTR_UP.update();
+      INSTR_DOWN.update();
+      INSTR_MOUSE.update();
+  }
+} // gameArea
+
+/* Returns true if the current frame number
+   corresponds with the given interval */
+function everyInterval(givenInterval) {
+  if ((gameArea.frameNo / givenInterval) % 1 == 0) {return true;}
+  return false;
+}
+
+var player;
+var obstacles = [];
+var score;
+
+/* multiple intruction boxes becaise
+   canvas does not support multi-line */
+var INSTR_UP;
+var INSTR_DOWN;
+var INSTR_MOUSE;
+
+function startGame() {
+  player = new component(30, 30, "red", 10, window.innerHeight/2);
+  score = new component("30px", "Consolas", "black", window.innerWidth - 200, 30, "text");
+
+  // create a component for each instruction box
+  INSTR_UP = new component("20px", "Consolas", "black", window.innerWidth - 250, window.innerHeight - 90, "text");
+  INSTR_DOWN = new component("20px", "Consolas", "black", window.innerWidth - 250, window.innerHeight - 65, "text");
+  INSTR_MOUSE = new component("20px", "Consolas", "black", window.innerWidth - 250, window.innerHeight - 40, "text");
+
+  // unicodes for down and up arrows
+  UPArrSymbol = "\u2B06";
+  DOWNArrSymbol = "\u2B07";
+
+  INSTR_UP.text = "w/" + UPArrSymbol + " > up";
+  INSTR_DOWN.text = "s/" + DOWNArrSymbol + " > down";
+  INSTR_MOUSE.text = "m > on/off mouse control";
+
+  gameArea.start();
+}
+
+/* Construct a component with specifiec width
+   height, color and coordinates where it is
+   places in the canvas */
+function component(width, height, color, x, y, type) {
+    this.type = type;
+    this.width = width;
+    this.height = height;
+    this.speedX = 0;
+    this.speedY = 0;
+    this.x = x;
+    this.y = y;
+    this.mouseControl = false;
+    this.update = function() {
+      ctx = gameArea.context;
+      if (this.type == "text") {
+        ctx.font = this.width + " " + this.height;
```

Comment:
This is a sign that the design of `component` can be improved. The parameters are `width` and `height`, but you're using them here as the font type, and you're concatenating them as a string to match the font spec. And the `type` parameter is just used to change how the component is updated. This will be hard to read in the future (I spent some time searching for the different `component` functions until I saw that you are reusing the parameters like this) and also error-prone.

The fact that everything is a `component` points to some kind of inheritance (you could have a base class Component with shared data/methods, and then subclasses TextComponent, BoxComponent, that implement the different behaviors).

However full inheritance here might be overkill, as there isn't much common data between the text components and the others - I think there's only the fact that both have an `update()` method? I think we can improve on this design in different ways - apart from the full-on inheritance:

- Because you might not need to call `update()` on the text components at every frame, you could just remove all text-related functionalities from `component`, and have a separate function that initializes the text with instructions at load time. (I made another comment about this - I'm unsure if you need to call `update` on every frame or not). Then, components are only used for things that can crash, or that move at every frame.

- We can keep everything as a `component` but we use different constructors to make the code more readable. You can have several constructor functions that will construct components for you, with descriptive names, and they make sure to create a component and set the `update` function appropriately. Then your function `component` becomes the "private" constructor, and you use exclusively the other two:

```
  // This is your previous `component` function - note how it doesn't take a type anymore.
  // This is not meant to be called anymore, outside of TextComponent and BoxComponent.
  function initializeComponent(width, height, color, x, y) {  
    // ...
  }

  function TextComponent(text, font, color, x, y) {
    initializeComponent(null, null, color, x, y);
    this.update = function() {
      ctx = gameArea.context;
      ctx.font = font;
      ctx.fillStyle = color;
      ctx.fillText(this.text, this.x, this.y);
    }
  }

  function BoxComponent(width, height, color, x, y) {
    initializeComponent(width, height, color, x, y);
    this.update = function() {
      ctx.fillStyle = color;
      ctx.fillRect(this.x, this.y, this.width, this.height);
    }
  }

  // Now you can do, e.g.
  player = new BoxComponent(30, 30, "red", 10, window.innerHeight/2);
  INSTR_UP = new TextComponent("w/" + UPArrSymbol + " > up", "20px Consolas", "black", window.innerWidth - 250, window.innerHeight - 90);
```

In this example, in TextComponent() I'm passing `null,null` as parameters which is another sign that the design can be further improved. We can remove them as parameters, and have BoxComponent() explicitly set `this.width = width`, and `this.height = height`. But in that case, there will be a disconnect, as the function crashWith inside initializeComponent is reading from this.width and this.height, and it's unclear where these fields are set up - at the same time, someone could call `var component = new initializeComponent(...)`, and then crashWith would run without having initialised those values.

- You can also remove the inheritance altogether and only two functions, BoxComponent and TextComponent. BoxComponent will contain the functions `update`, `newPos`, `newSpeed`, and `crashWith`. TextComponent will contain its own `update`. This is again not ideal :) as now there isn't any mechanism that ensures that both components have an `update` function.

This was a long comment - I'm happy to discuss more about other alternatives over chat or at a meeting. I'll find the proper way of doing this in Javascript using classes/prototype inheritance. For now, any of these solutions is good - as long as we make the code more readable and a bit less error-prone.

Reply:
Lack of readability was one of my concerns too when I wrote the code and realised that the design could be improved significantly. I opted for the last solution that you mentioned, namely I have two separate constructors for the box and text components with their respective functions and attributes. The only common function is the update function.